zebra-network/src/meta_addr.rs

```diff
@@ -259,18 +438,208 @@ impl MetaAddr {
+            // only put the last seen time in the untrusted field,
+            // this matches deserialization, and avoids leaking internal state
+            untrusted_last_seen: Some(last_seen),
+            last_response: None,
+            // these fields aren't sent to the remote peer, but sanitize them anyway
+            last_attempt: None,
+            last_failure: None,
             last_connection_state: NeverAttemptedGossiped,
+        })
+    }
+}
+
+impl MetaAddrChange {
+    /// Return the address for this change.
+    pub fn addr(&self) -> SocketAddr {
+        match self {
+            NewGossiped { addr, .. }
+            | NewAlternate { addr, .. }
+            | NewLocal { addr, .. }
+            | UpdateAttempt { addr }
+            | UpdateResponded { addr, .. }
+            | UpdateFailed { addr, .. } => *addr,
+        }
+    }
+
+    #[cfg(any(test, feature = "proptest-impl"))]
+    /// Set the address for this change to `new_addr`.
+    ///
+    /// This method should only be used in tests.
+    pub fn set_addr(&mut self, new_addr: SocketAddr) {
+        match self {
+            NewGossiped { addr, .. }
+            | NewAlternate { addr, .. }
+            | NewLocal { addr, .. }
+            | UpdateAttempt { addr }
+            | UpdateResponded { addr, .. }
+            | UpdateFailed { addr, .. } => *addr = new_addr,
+        }
+    }
+
+    /// Return the untrusted services for this change, if available.
+    pub fn untrusted_services(&self) -> Option<PeerServices> {
+        match self {
+            NewGossiped {
+                untrusted_services, ..
+            } => Some(*untrusted_services),
+            NewAlternate {
+                untrusted_services, ..
+            } => Some(*untrusted_services),
+            // TODO: create a "services implemented by Zebra" constant
+            NewLocal { .. } => Some(PeerServices::NODE_NETWORK),
+            UpdateAttempt { .. } => None,
+            UpdateResponded { services, .. } => Some(*services),
+            UpdateFailed { services, .. } => *services,
+        }
+    }
+
+    /// Return the untrusted last seen time for this change, if available.
+    pub fn untrusted_last_seen(&self) -> Option<DateTime32> {
+        match self {
+            NewGossiped {
+                untrusted_last_seen,
+                ..
+            } => Some(*untrusted_last_seen),
+            NewAlternate { .. } => None,
+            NewLocal { .. } => None,
+            UpdateAttempt { .. } => None,
+            UpdateResponded { .. } => None,
+            UpdateFailed { .. } => None,
+        }
+    }
+
+    /// Return the last attempt for this change, if available.
+    pub fn last_attempt(&self) -> Option<Instant> {
+        match self {
+            NewGossiped { .. } => None,
+            NewAlternate { .. } => None,
+            NewLocal { .. } => None,
+            UpdateAttempt { .. } => Some(Instant::now()),
```

Comment:
It's not an issue in practice. But the impacts are different for each time field, and some of the reasoning is subtle.

I've added detailed comments in ccdb6a81 in #2275.

I also updated #1976 so the refactor updates the attempt time before we start the handshake. (Otherwise we could try the same peer twice if there are channel delays.)

As a side-effect, keeping times out of changes makes them smaller and more efficient. It also allows us to optimise timestamp collection by merging identical changes, see #2293.